heat/decomposition/pca.py

```diff
@@ -0,0 +1,278 @@
+        self.singular_values_ = None
+        self.mean_ = None
+        self.n_components_ = None
+        self.n_samples_ = None
+        self.noise_variance_ = None
+
+    def fit(self, X: ht.DNDarray, y=None) -> Self:
+        """
+        Fit the PCA model with data X.
+
+        Parameters
+        ----------
+        X : DNDarray of shape (n_samples, n_features)
+            Data set of which PCA has to be computed.
+        y : Ignored
+            Not used, present for API consistency by convention.
+        """
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("X must be a DNDarray.")
+        if y is not None:
+            raise ValueError(
+                "Argument y is ignored and just present for API consistency by convention."
+            )
+
+        # center data
+        self.mean_ = X.mean(axis=0)
+        X_centered = X - self.mean_
+
+        # set n_components
+        # note: n_components is an argument passed by the user to the PCA object and encodes the TARGET number of components
+        #       n_components_ is an attribute of the PCA object that stores the ACTUAL number of components
+        if self.n_components is None:
+            self.n_components_ = min(X.shape)
+        else:
+            self.n_components_ = self.n_components
+
+        # compute SVD via "full" SVD
+        if self.svd_solver == "full" or not X.is_distributed():
+            _, S, V = ht.linalg.svd(X_centered, full_matrices=False)
+            if isinstance(self.n_components_, int):
+                # prescribed truncation rank (including no truncation)
+                self.components_ = V[:, : self.n_components_].T
+                self.singular_values_ = S[: self.n_components_]
+
+                self.explained_variance_ = (S**2)[: self.n_components_] / (X.shape[0] - 1)
+                total_variance = (S**2).sum() / (X.shape[0] - 1)
+                self.explained_variance_ratio_ = self.explained_variance_ / total_variance
+
+            else:
+                # truncation w.r.t. prescribed bound on explained variance
+                total_variance = (S**2).sum() / (X.shape[0] - 1)
+                explained_variance_threshold = self.n_components_ * total_variance.larray.item()
+                explained_variance_cumsum = (S**2).larray.cumsum(0) / (X.shape[0] - 1)
+                self.n_components_ = (
+                    len(
+                        explained_variance_cumsum[
+                            explained_variance_cumsum <= explained_variance_threshold
+                        ]
+                    )
+                    + 1
+                )
+                self.components_ = V[:, : self.n_components_].T
+                self.singular_values_ = S[: self.n_components_]
+                self.explained_variance_ = (S**2)[: self.n_components_] / (X.shape[0] - 1)
+                self.explained_variance_ratio_ = self.explained_variance_ / total_variance
```

Comment:
This can be refactored. The blocks are mostly the same.

Reply:
good hint! thx :+1: